plugins/woocommerce/src/Internal/DependencyManagement/RuntimeContainer.php

```diff
@@ -0,0 +1,225 @@
+	 * @throws ContainerException Error when resolving the class to an object instance.
+	 */
+	protected function get_core( string $class_name, array &$resolve_chain ) {
+		if ( isset( $this->resolved_cache[ $class_name ] ) ) {
+			return $this->resolved_cache[ $class_name ];
+		}
+
+		// phpcs:disable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		if ( in_array( $class_name, $resolve_chain, true ) ) {
+			throw new ContainerException( "Recursive resolution of class '$class_name'. Resolution chain: " . implode( ', ', $resolve_chain ) );
+		}
+
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which is not in the " . self::WOOCOMMERCE_NAMESPACE . ' namespace. Did you forget to add a namespace import?' );
+		}
+
+		if ( ! class_exists( $class_name ) ) {
+			throw new ContainerException( "Attempt to get an instance of class '$class_name', which doesn't exist." );
+		}
+
+		// Account for the containers used by the Store API and Blocks.
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\StoreApi\\' ) ) {
+			return StoreApi::container()->get( $class_name );
+		}
+		if ( StringUtil::starts_with( $class_name, 'Automattic\WooCommerce\Blocks\\' ) ) {
+			return BlocksPackage::container()->get( $class_name );
+		}
+
+		$resolve_chain[] = $class_name;
+
+		try {
+			$instance = $this->instantiate_class_using_reflection( $class_name, $resolve_chain );
+		} catch ( \ReflectionException $e ) {
+			throw new ContainerException( "Reflection error when resolving '$class_name': (" . get_class( $e ) . ") {$e->getMessage()}", 0, $e );
+		}
+
+		// phpcs:enable WordPress.Security.EscapeOutput.ExceptionNotEscaped
+
+		$this->resolved_cache[ $class_name ] = $instance;
+
+		return $instance;
+	}
+
+	/**
+	 * Get an instance of a class using reflection.
+	 * This method recursively calls 'get_core' (which in turn calls this method) for each of the arguments
+	 * in the 'init' method of the resolved class (if the method is public and non-static).
+	 *
+	 * @param string $class_name The name of the class to resolve.
+	 * @param array  $resolve_chain Classes already resolved in this resolution chain. Passed between recursive calls to the method in order to detect a recursive resolution condition.
+	 * @return object The resolved object.
+	 *
+	 * @throws ContainerException The 'init' method has invalid arguments.
+	 * @throws \ReflectionException Something went wrong when using reflection to get information about the class to resolve.
+	 */
+	private function instantiate_class_using_reflection( string $class_name, array &$resolve_chain ): object {
+		$ref_class = new \ReflectionClass( $class_name );
+		$instance  = $ref_class->newInstance();
+		if ( ! $ref_class->hasMethod( 'init' ) ) {
+			return $instance;
+		}
+
+		$init_method = $ref_class->getMethod( 'init' );
+		if ( ! $init_method->isPublic() || $init_method->isStatic() ) {
```

Comment:
Should we require only `final` injection methods, as before? 🤔

Reply:
I think having the check in the linter is enough, but I'm open to discussion if there's any tangible benefit on enforcing `final`-ness at runtime.

Reply:
Honestly, I don’t see much benefit to enforcing `final` beyond consistency. The linter should handle it, so I’m good with keeping it as-is if that works for everyone.